scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,306 @@
+                (assoc :valid? false
+                       :reason (str unique-constraint-field " is not unique between clusters " keys-for-unique-value))))))))
+
+(defn compute-config-updates
+  "Take the current and desired configurations and compute the changes. Alert on invalid changes."
+  [db current-configs new-configs force?]
+  (let [[deletes-keys inserts-keys updates-keys] (diff-map-keys current-configs new-configs)
+        changes (->> (concat
+                       (map #(let [current (current-configs %)]
+                               (compute-config-update db current (assoc current :state :deleted) force?)) deletes-keys)
+                       (map #(compute-config-insert (new-configs %)) inserts-keys)
+                       (map #(compute-config-update db (current-configs %) (new-configs %) force?) updates-keys))
+                     (filter :changed?))
+        resulting-configs (->> changes
+                               (filter :valid?)
+                               (#(for-map [change %] (:cluster-name change) (:config change)))
+                               (merge current-configs))
+        ;TODO check that number of running clusters is not less than a configured value. and make no changes if so
+        ]
+    (-> changes
+      (check-for-unique-constraint-violations resulting-configs :base-path)
+      (check-for-unique-constraint-violations resulting-configs :ca-cert))))
+
+;TODO see if this is ok or need a better way
+(def scheduler-promise (promise))
+(def exit-code-syncer-state-promise (promise))
+
+(defn add-new-cluster!
+  "Add a new cluster from a dynamic cluster config"
+  [conn config]
+  (try
+    (let [config-from-template ((config/compute-cluster-templates) (:template config))
+          factory-fn (:factory-fn config-from-template)
+          resolved (cook.util/lazy-load-var factory-fn)
+          config (merge (:config config-from-template) config)
+          cluster (resolved config {:exit-code-syncer-state @exit-code-syncer-state-promise})]
+      (initialize-cluster cluster (:pool-name->fenzo @scheduler-promise)))
+    (let [temp-db-id (d/tempid :db.part/user)]
+      @(d/transact
+         conn
+         [(assoc (compute-cluster-config->compute-cluster-config-ent config) :db/id temp-db-id)]))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+(defn update-cluster!
+  "Update a cluster with a dynamic cluster config"
+  [conn {:keys [name state state-locked?] :as config} current-db-config-ents current-in-mem-configs]
+  (try
+    (let [{:keys [state-atom state-locked?-atom] :as cluster} (@cluster-name->compute-cluster-atom name)
+          current-db-config-ent (current-db-config-ents name)
+          current-in-mem-config (current-in-mem-configs name)]
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
```

Comment:
This seems weird to me. Why can locked cause problems at this point? Shouldn't we just reject the REST API at submit, long before here?

Reply:
this shouldn't be possible. but the way I wrote the logic below this was implied and I just added the check. I'd probably be fine removing these

Reply:
Ok.

Reply:
updated